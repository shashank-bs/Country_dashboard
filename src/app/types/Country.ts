export interface Country {
    name: {
      common: string;
      official:string;
    };
    population: number;
    region: string;
    capital: string[];
    flags: {
      svg: string;
    };
    currencies: Record<string, { name: string }>;
    languages: Record<string, string>;
    timezones: string[];
    cca2:string;
  }
  