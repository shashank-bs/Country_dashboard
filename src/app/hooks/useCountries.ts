// hooks/useCountries.ts
import { useState, useEffect } from "react";
import axios from "axios";
import { Country } from "../types/Country";

const useCountries = () => {
  const [countries, setCountries] = useState<Country[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchCountries = async () => {
      setLoading(true);
      try {
        const response = await axios.get("https://restcountries.com/v3.1/all");
        setCountries(response.data);
      } catch (err) {
        setError("Failed to fetch countries");
      } finally {
        setLoading(false);
      }
    };
    fetchCountries();
  }, []);

  return { countries, loading, error };
};

export default useCountries;
