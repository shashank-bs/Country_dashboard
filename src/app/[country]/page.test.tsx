// src/app/[country]/page.test.tsx
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { useParams } from "next/navigation";
import CountryPage from "./page";
import useCountries from "../hooks/useCountries";
import { Country } from "../types/Country";

// Mock the useParams hook
jest.mock("next/navigation", () => ({
  useParams: jest.fn(),
}));

// Mock the useCountries hook
jest.mock("../hooks/useCountries", () => ({
  __esModule: true,
  default: jest.fn(),
}));

const mockCountry: Country = {
  name: {
    common: "United States",
    official: "",
  },
  flags: { svg: "https://example.com/flag.svg" },
  capital: ["Washington, D.C."],
  population: 331000000,
  region: "Americas",
  currencies: {},
  languages: {},
  timezones: [],
  cca2: "",
};

describe("CountryPage Component", () => {
  beforeEach(() => {
    (useParams as jest.Mock).mockReturnValue({ country: "United States" });
  });

  test("displays loading state initially", () => {
    (useCountries as jest.Mock).mockReturnValue({
      countries: [],
      loading: true,
    });
    render(<CountryPage />);
    expect(screen.getByRole("progressbar")).toBeInTheDocument();
  });

  test("displays 'Country not found' when country is not found", () => {
    (useCountries as jest.Mock).mockReturnValue({
      countries: [],
      loading: false,
    });
    render(<CountryPage />);
    expect(screen.getByText("Country not found")).toBeInTheDocument();
  });

  test("displays country details when country is found", () => {
    (useCountries as jest.Mock).mockReturnValue({
      countries: [mockCountry],
      loading: false,
    });
    render(<CountryPage />);
    expect(screen.getByText("United States")).toBeInTheDocument();
  });

  test("displays country details when country is found", () => {
    (useCountries as jest.Mock).mockReturnValue({
      countries: [mockCountry],
      loading: false,
    });
    render(<CountryPage />);
    expect(screen.getByText("United States")).toBeInTheDocument();
  });

  test("matches snapshot when loading", () => {
    (useCountries as jest.Mock).mockReturnValue({
      countries: [],
      loading: true,
    });
    const { asFragment } = render(<CountryPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot when country is not found", () => {
    (useCountries as jest.Mock).mockReturnValue({
      countries: [],
      loading: false,
    });
    const { asFragment } = render(<CountryPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot when country is found", () => {
    (useCountries as jest.Mock).mockReturnValue({
      countries: [mockCountry],
      loading: false,
    });
    const { asFragment } = render(<CountryPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
