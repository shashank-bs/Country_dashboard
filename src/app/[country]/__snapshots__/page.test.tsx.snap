// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryPage Component matches snapshot when country is found 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-1wcaknn"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 css-10zsxzp-MuiTypography-root"
      >
        United States
      </h4>
      <h6
        class="MuiTypography-root MuiTypography-h6 css-8zgow4-MuiTypography-root"
      >
        Capital: Washington, D.C.
      </h6>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
      >
        Population: 33,10,00,000
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
      >
        Region: Americas
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
      >
        Timezones: 
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
      >
        Languages: 
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
      >
        Currencies: 
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CountryPage Component matches snapshot when country is not found 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-14h3iqr"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 css-1vcvfao-MuiTypography-root"
    >
      Country not found
    </h5>
  </div>
</DocumentFragment>
`;

exports[`CountryPage Component matches snapshot when loading 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-14h3iqr"
  >
    <span
      class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1yonwsg-MuiCircularProgress-root"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-1a8epmm-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;
